x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -109,6 +109,55 @@ ResolvedIndices resolve(String action, TransportRequest request, Metadata metada
         return resolveIndicesAndAliases(action, (IndicesRequest) request, metadata, authorizedIndices);
     }
 
+    boolean requiresWildcardExpansion(IndicesRequest indicesRequest) {
+        if (indicesRequest instanceof IndicesAliasesRequest) {
+            return true;
+        } else {
+            return indicesRequest instanceof IndicesRequest.Replaceable;
+        }
+    }
+
+    ResolvedIndices resolveIndicesAndAliases(String action, IndicesRequest indicesRequest) {
+        assert false == requiresWildcardExpansion(indicesRequest) : "indices must not require wildcard expansion";
+        final String[] indices = indicesRequest.indices();
+        if (indices == null || indices.length == 0) {
+            throw new IllegalArgumentException("the action " + action + " requires explicit index names, but none were provided");
+        }
+        if (IndexNameExpressionResolver.isAllIndices(Arrays.asList(indices))) {
+            throw new IllegalArgumentException(
+                "the action "
+                    + action
+                    + " does not support accessing all indices;"
+                    + " the provided index expression ["
+                    + Strings.arrayToCommaDelimitedString(indices)
+                    + "] is not allowed"
+            );
+        }
+
+        // TODO: Shard level requests have wildcard expanded already and do not need go through this check
+        final List<String> wildcards = Stream.of(indices).filter(Regex::isSimpleMatchPattern).collect(Collectors.toList());
+        if (wildcards.isEmpty() == false) {
+            throw new IllegalArgumentException(
```

Comment:
Requests like `ShardSearchRequest` already has its wildcard expanded and does not need this check. We can gain some performance here for large list of indices if we have an allowlist to skip this check. 

TransportClient can potentially send a ShardLevelRequest directly with wildcard not expanded. But I don't think we need to worry about it since the actual search will not rely on the indices (it uses ShardSearchRequest#shardId). Also the comment below clearly states that it is acceptable to allow wildcards in shard level requests. They just won't get expanded.

It would be great if we have a formal way to categorize non-replaceable requests into (1) directly created for REST level requests (e.g. `PutIndexRequest`) and (2) *NOT* directly created for REST level requests (e.g. `ShardSearchRequest`). Because major part of this method is necessary for (1) but not (2). For now, an allowlist is an viable alternative.